oss-internship-2020/libtiff/test/raw_decode.cc

```diff
@@ -0,0 +1,199 @@
+      << ")";
+  return comp;
+}
+
+int CheckRgbaPixel(int pixel, int min_red, int max_red, int min_green,
+                   int max_green, int min_blue, int max_blue, int min_alpha,
+                   int max_alpha, const sapi::v::Array<unsigned>& buffer) {
+  // RGBA images are upside down - adjust for normal ordering
+  int adjusted_pixel = pixel % 128 + (127 - (pixel / 128)) * 128;
+  unsigned rgba = buffer[adjusted_pixel];
+
+  bool comp = !(TIFFGetR(rgba) >= (unsigned)min_red &&
+                TIFFGetR(rgba) <= (unsigned)max_red &&
+                TIFFGetG(rgba) >= (unsigned)min_green &&
+                TIFFGetG(rgba) <= (unsigned)max_green &&
+                TIFFGetB(rgba) >= (unsigned)min_blue &&
+                TIFFGetB(rgba) <= (unsigned)max_blue &&
+                TIFFGetA(rgba) >= (unsigned)min_alpha &&
+                TIFFGetA(rgba) <= (unsigned)max_alpha);
+
+  EXPECT_THAT(comp, IsFalse())
+      << "Pixel " << pixel << " did not match expected results.\n"
+      << "Got R=" << TIFFGetR(rgba) << " (expected " << min_red << ".."
+      << max_red << "), G=" << TIFFGetG(rgba) << " (expected " << min_green
+      << ".." << max_green << "), B=" << TIFFGetB(rgba) << " (expected "
+      << min_blue << ".." << max_blue << "), A=" << TIFFGetA(rgba)
+      << " (expected " << min_alpha << ".." << max_alpha << ")";
+  return comp;
+}
+
+TEST(SandboxTest, RawDecode) {
+  std::string srcfile = GetFilePath("quad-tile.jpg.tiff");
+
+  TiffSapiSandbox sandbox("", srcfile);
+  ASSERT_THAT(sandbox.Init(), IsOk()) << "Couldn't initialize Sandboxed API";
+
+  tsize_t sz;
+  unsigned int pixel_status = 0;
+  sapi::v::UShort h, v;
+  sapi::StatusOr<TIFF*> status_or_tif;
+  sapi::StatusOr<int> status_or_int;
+  sapi::StatusOr<tmsize_t> status_or_long;
+
+  TiffApi api(&sandbox);
+  sapi::v::ConstCStr srcfile_var(srcfile.c_str());
+  sapi::v::ConstCStr r_var("r");
+
+  status_or_tif = api.TIFFOpen(srcfile_var.PtrBefore(), r_var.PtrBefore());
+  ASSERT_THAT(status_or_tif, IsOk()) << "Could not open " << srcfile;
+
+  sapi::v::RemotePtr tif(status_or_tif.value());
+  ASSERT_THAT(tif.GetValue(), NotNull())
+      << "Could not open " << srcfile << ", TIFFOpen return NULL";
+
+  status_or_int = api.TIFFGetField2(&tif, TIFFTAG_YCBCRSUBSAMPLING, h.PtrBoth(),
+                                    v.PtrBoth());
+  ASSERT_THAT(status_or_int, IsOk()) << "TIFFGetField2 fatal error";
+  EXPECT_THAT(
+      status_or_int.value() == 0 || h.GetValue() != 2 || v.GetValue() != 2,
+      IsFalse())
+      << "Could not retrieve subsampling tag";
+
+  status_or_long = api.TIFFTileSize(&tif);
+  ASSERT_THAT(status_or_int, IsOk()) << "TIFFTileSize fatal error";
+  EXPECT_THAT(status_or_long.value(), Eq(24576))
+      << "tiles are " << status_or_long.value() << " bytes";
+  sz = status_or_long.value();
+
+  sapi::v::Array<uint8_t> buffer_(sz);
+  status_or_long = api.TIFFReadEncodedTile(&tif, 9, buffer_.PtrBoth(), sz);
```

Comment:
Magic 9 as well - I will stop commenting from here on on things that I've already mentioned before